attestation-agent/attester/src/tsm_report/mod.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2024 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use log::error;
+use std::path::{Path, PathBuf};
+use tempfile::tempdir_in;
+use thiserror::Error;
+
+const TSM_REPORT_PATH: &str = "/sys/kernel/config/tsm/report";
+pub const TSM_REPORT_PROVIDER_TDX: &str = "tdx_guest";
+pub const TSM_REPORT_PROVIDER_SEV: &str = "sev_guest";
+
+#[derive(Error, Debug)]
+pub enum TsmReportError {
+    #[error("Failed to create TSM Report path: {0}")]
+    Open(#[from] std::io::Error),
+    #[error("Failed to access TSM Report attribute: {0}: {1}")]
+    Access(&'static str, #[source] std::io::Error),
+    #[error("TSM Report inblob write conflict: generation {0}, expected 1")]
+    InblobConflict(u32),
+    #[error("Failed to generate TSM Report: missing inblob (len=0)")]
+    InblobLen,
+}
+
+pub enum TsmReportProvider {
+    Tdx(Vec<u8>),
+    Sev(u8, Vec<u8>),
+}
+
+pub struct TsmReportPath {
+    path: PathBuf,
+    pub provider: String,
+}
+
+impl TsmReportPath {
+    pub fn open() -> Result<Self, TsmReportError> {
+        let p = tempdir_in(TSM_REPORT_PATH).map_err(TsmReportError::Open)?;
+
+        let path = p.into_path();
+
+        let provider = match std::fs::read_to_string(path.as_path().join("provider")) {
+            Ok(tsm) => tsm.trim_matches('\n').to_string(),
+            Err(e) => {
+                let _ = std::fs::remove_dir(path.as_path())
+                    .map_err(|r| log::error!("Failed to remove TSM Report directory: {}", r));
+                return Err(TsmReportError::Access("provider", e));
+            }
+        };
+
+        Ok(Self { path, provider })
+    }
+    pub fn close(&self) {
+        let _ = std::fs::remove_dir(self.path.as_path())
+            .map_err(|e| log::error!("Failed to remove TSM Report directory: {}", e));
+    }
+    pub fn attestation_report(
+        &self,
+        provider: TsmReportProvider,
+    ) -> Result<Vec<u8>, TsmReportError> {
+        let report_path = self.path.as_path();
+
+        let report_data = match provider {
+            TsmReportProvider::Tdx(inblob) => inblob,
+            TsmReportProvider::Sev(privlevel, inblob) => {
```

Comment:
I found that there is a logic here
- if `attester-tdx` is enabled, check either TSM Report or ioctl dev is supported
- if succeeds, try TSM Report first and ioctl if fails

However, TSM supports both TDX and SNP, so would it be
- if `tsm-report` is detected, go into `tsm-report` submodule and detect which provider is
- use match branch to provider in blob and then get out blob
- else, go legacy routine, s.t. detect ioctl TDX, SNP, ...

Reply:
> However, TSM supports both TDX and SNP, so would it be

the logic for other implementations would follow the TDX example: sev attester uses `provider_is(SEV`)  to check TSM reports for SEV are available. sev attesters `get_evidence` implements it's own logic for getting the quote using TSM reports (`Sev()` enum + attestation_report + supplemental_data)